src/imputors/knn.jl

```diff
@@ -17,59 +17,77 @@ struct KNN{M} <: Imputor where M <: NearestNeighbors.MinkowskiMetric
-        mmask = ismissing.(transpose(data))
-
-        # fill missing value as mean value
-        impute!(data, Fill(; value=mean, context=ctx))
-
-        # then, transpose to D x N for KDTree
-        transposed = transpose(disallowmissing(data))
-
-        kdtree = KDTree(transposed, imp.dist)
-        idxs, dists = NearestNeighbors.knn(kdtree, transposed, imp.k, true)
-
-        idxes = CartesianIndices(transposed)
-        fallback_threshold = imp.k * imp.threshold
-
-        for I in CartesianIndices(transposed)
-            if mmask[I] == 1
-                w = 1.0 ./ dists[I[2]]
-                ws = sum(w[2:end])
-                missing_neighbors = ismissing.(transposed[:, idxs[I[2]]][:, 2:end])
-
-                # exclude missing value itself because distance would be zero
-                if isnan(ws) || isinf(ws) || iszero(ws)
-                    # if distance is zero or not a number, keep mean imputation
-                    transposed[I] = transposed[I]
-                elseif count(!iszero, mapslices(sum, missing_neighbors, dims=1)) >
-                    fallback_threshold
-                    # If too many neighbors are also missing, fallback to mean imputation
-                    # get column and check how many neighbors are also missing
-                    transposed[I] = transposed[I]
-                else
-                    # Inverse distance weighting
-                    wt = w .* transposed[I[1], idxs[I[2]]]
-                    transposed[I] = sum(wt[2:end]) / ws
-                end
+function impute!(data::AbstractMatrix{Union{T, Missing}}, imp::KNN; dims=nothing) where T<:Real
+    if dims === nothing
+        Base.depwarn(
+            "Imputing on matrices will require specifying `dims=2` in a future release, " *
+            "to maintain the current behaviour.",
+            :impute!
+        )
+        dims = 2
+    elseif dims === :rows
+        dims = 1
+    elseif dims === :cols
+        dims = 2
+    end
+
+    # KDTree expects data of the form dims x n
+    X = dims == 1 ? data : transpose(data)
+
+    # Get mask array first
+    mmask = ismissing.(X)
+
+    # fill missing value as mean value
+    impute!(X, Fill(; value=mean); dims=:rows)
+
+    # Disallow `missings` for NearestNeighbors
+    X = disallowmissing(X)
+
+    kdtree = KDTree(X, imp.dist)
+    idxs, dists = NearestNeighbors.knn(kdtree, X, imp.k, true)
+
+    idxes = CartesianIndices(X)
+    fallback_threshold = imp.k * imp.threshold
+
+    for I in CartesianIndices(X)
+        if mmask[I] == 1
+            w = 1.0 ./ dists[I[2]]
+            ws = sum(w[2:end])
+            # Shouldn't ismissing.(X[...][...]) be replaced with mmask[...][...]?
+            # If so then I think the test might need updating cause the "Data match" section
+            # seems to fallback on the mean imputation consistently
+            neighbors = mapslices(
+                iszero ∘ sum,
+                ismissing.(X[:, idxs[I[2]]][:, 2:end]);
```

Comment:
@appleparan Should this be `mmask[:, idxs[I[2]]][:, 2:end]`? AFAICT, this will always be an array of falses because `X` (previously `transposed`) is already `mean` imputed.

Reply:
Made an issue for this for now. https://github.com/invenia/Impute.jl/issues/73